assets/app/view/game/buy_trains.rb

```diff
@@ -54,6 +54,8 @@ def render_president_contributions
       def render
         step = @game.round.active_step
         @corporation = step.current_entity
+        @entity = @selected_company&.owner == @corporation ? @selected_company : @corporation
```

Comment:
you should only set the entity if the company has train buying abilities, wouldn't the current code allow any company to buy a train?

Reply:
Have rewritten this part. Now only set @ability here and do an explicit use of entity where it is needed.